shoop/admin/modules/orders/json_order_creator.py

```diff
@@ -277,6 +276,22 @@ def create_source_from_state(self, state, creator=None, ip_address=None,
         if not self.is_valid:
             return None
 
+        if order_to_update:
+            # Preserve values that is not in state from original order
+            source.update(
+                creator=order_to_update.creator,
+                ip_address=order_to_update.ip_address,
+                orderer=order_to_update.orderer,
+                order_date=order_to_update.order_date,
+                status=order_to_update.status,
+                marketing_permission=order_to_update.marketing_permission,
+                customer_comment=order_to_update.customer_comment,
+                payment_data=(order_to_update.payment_data or {}),
+                shipping_data=(order_to_update.shipping_data or {}),
+                extra_data=(order_to_update.extra_data or {}),
+            )
+            for code in order_to_update.coupon_usages.values_list("coupon__code", flat=True):
+                source.add_code(code)
```

Comment:
You shouldn't set codes by coupon usages. Coupon usages is different as the codes in order_source. Coupons are something that shoop.campaigns uses, but order_source.codes is more general concept of shoop.core. It might also very well be that shoop.campaigns isn't enabled at all in INSTALLED_APPS and then there even isn't `order.coupon_usages` at all and your code would fail.

You should rather create a new field to order which stores the codes from order_source and then just use that to repopulate the order_source.codes, e.g. like this.

```python
for code in order_to_update.codes:
    source.add_code(code)
```

and in order_creator, do:

```python
order.codes = order_source.codes
```

`Order.codes` should probably be a property to protect it being accessed directly, like this:
```python
class Order(...):
    ...

    _codes = JSONField(...)

    @property
    def codes():
        return list(self._codes or [])

    @codes.setter
    def codes(self, value):
        codes = []
        for code in value:
            if not isinstance(code, six.text_type):
                raise TypeError('codes must be a list of strings')
            codes.append(code)
        self._codes = codes
```